### ISSUE TITLE
Logging drivers enhancement 

### ISSUE THREAD

**Issue by fahedouch** [containerd/nerdctl#1039]
### What is the problem you're trying to solve

Target Nerdctl logging drivers:

- [ ] none
- [x] json-file : https://github.com/containerd/nerdctl/pull/959
- [ ] local
- [x] journald : https://github.com/containerd/nerdctl/pull/1062
- [x] fluentd : https://github.com/containerd/nerdctl/pull/1073
- [ ] fluentbit
- [ ] syslog

### Describe the solution you'd like

Configure logging drivers

**Comment by Zheaoli** [comment]
Maybe I can handle the rest of the task in this issue. @fahedouch would you mind this issue to me?

**Comment by fahedouch** [comment]
thank you for  you request  @Zheaoli . I start writing code for `journald` so let me finish it based on your refacto https://github.com/containerd/nerdctl/pull/1050 . You can enhance the rest of the list

**Comment by Zheaoli** [comment]
@fahedouch Great! Maybe I can write the `fluentd` and `syslog` first. BTW, maybe we need spilt the logger in two parts: the readable logger and the unreadable logger

**Comment by fahedouch** [comment]
@Zheaoli  What do you mean by readable logger and the unreadable logger here ?

**Comment by Zheaoli** [comment]
Yes, because some of the loggers such as `fluentd` can not read the log directly after we wrote the log. So I think maybe we need to split the logger by two types

**Comment by AkihiroSuda** [comment]
Probably we should support using multiple drivers simultaneously. Then we can read/write jsonfiles while writing to fluents.

**Comment by Zheaoli** [comment]
> Probably we should support using multiple drivers simultaneously. Then we can read/write jsonfiles while writing to fluentd.

SGTM.

**Comment by fahedouch** [comment]
> Probably we should support using multiple drivers simultaneously. Then we can read/write jsonfiles while writing to fluentd.

We are creating a dependency between drivers here; `json-file` driver issue will impact `fluentd` driver from nerdctl point of view. This can be voluntary when the user configures jsonfiles as fluentd  `tail` input (https://docs.fluentd.org/input/tail) 
I think we should implement `fluentd` driver in a separated way ( we can implement one of the fluentd input https://docs.fluentd.org/input/tcp )

**Comment by fahedouch** [comment]
@Zheaoli FYI I start working on `local` driver.

**Comment by Zheaoli** [comment]
> @Zheaoli implement the above list is enough at least for the moment

SGTM

**Comment by AkihiroSuda** [comment]
I think the AWS CloudWatch driver should be implemented as a third party plugin that can be called with --log-driver binary://foo

**Comment by Zheaoli** [comment]
> I think the AWS CloudWatch driver should be implemented as a third party plugin that can be called with --log-driver binary://foo

SGTM, I think I can make a design proposal

**Comment by manugupt1** [comment]
I would like to try working on fluentbit driver if that is fine and no one is working on it.

**Comment by AkihiroSuda** [comment]
@containerd/nydus-maintainers 

Are we fine to add the fluentbit driver as a built-in driver, or should it be an external driver (--log-driver binary://foo) ?

**Comment by fahedouch** [comment]
@AkihiroSuda I would say external driver;  fluentbit driver(s) are be built as go plugin https://docs.fluentbit.io/manual/development/golang-output-plugins#build-a-go-plugin which can be easily handled as external driver. Also fluentbit have multiple deps so I think it is better to isolate it as external plugin. 

@manugupt1  Yes you can start working on this. I started implementing the local driver which also an external plugin. We just have to agree on the structuring of external plugins

**Comment by manugupt1** [comment]
@fahedouch do you have a draft PR for local that I can take inspiration from for external driver?

**Comment by Zheaoli** [comment]
@manugupt1 is there any update about this issue?

**Comment by manugupt1** [comment]
> @manugupt1 is there any update about this issue?

Sorry, been busy with work a lot. So, I have not made much progress

**Comment by manugupt1** [comment]
@AkihiroSuda @fahedouch @Zheaoli Should we close this as third party support is open or keep it open for inbuilt drivers?

**Comment by dalvarezquiroga** [comment]
Yeah would be great to read the logs using directly nerdctl :-)

nerdctl -n k8s.io container logs e780b0e8aba7
WARN[0000] Currently, `nerdctl logs` only supports containers created with `nerdctl run -d`
FATAL[0000] open /var/lib/nerdctl/1935db59/containers/e780b0e8aba78200dfe6112dc834925952ee6bb3a39f64cb8ab88081c77c74f9/log-config.json: no such file or directory

**Comment by anonymousr007** [comment]
Hello Everyone,

I am Rishabh, an undergraduate computer engineering student. I am interested in this issue. Is it still open?

Best Regards
Rishabh Singh

**Comment by fahedouch** [comment]
> Hello Everyone,
> 
> I am Rishabh, an undergraduate computer engineering student. I am interested in this issue. Is it still open?
> 
> Best Regards Rishabh Singh

Hi Rishabh,
Yes, please go ahead. What lnterest you in this ticket ?

**Comment by anonymousr007** [comment]
> > Hello Everyone,
> > I am Rishabh, an undergraduate computer engineering student. I am interested in this issue. Is it still open?
> > Best Regards Rishabh Singh
> 
> Hi Rishabh, Yes, please go ahead. What lnterest you in this ticket ?

- [ ] local
- [ ] fluentbit
- [ ] syslog

I guess, These three logging drivers are not implemented yet.

**Comment by austinvazquez** [comment]
Hi @fahedouch, can this ticket also include adding support for "none" logging driver?

The use case is to be able to disable container logs from compose file.

e.g. `nerdctl compose up` of `docker-compose.yml`
```
version: "3.7"

services:
  compute:
    image: public.ecr.aws/ubuntu/ubuntu:22.04
    logging:
      driver: "none"
```

results in:

```
INFO[0000] Ensuring image public.ecr.aws/ubuntu/ubuntu:22.04
INFO[0000] Re-creating container repro-compute-1
FATA[0000] unknown logging driver "none": not found
FATA[0000] error while creating container repro-compute-1: exit status 1
```

**Comment by vsiravar** [comment]
@fahedouch are you working on the local log driver? Do you have a draft for the design for inspiration :). I can help with the implementation.

**Comment by fahedouch** [comment]
@vsiravar  unfortunately I didn't have enough time to make a clear enough design. The idea is to separate the `local` driver into a different binary then consume it using `containerd runtime v2 logging plugin mode.`
```
"binary://BIN?KEY=VALUE" URI is parsed into Args {BIN, KEY, VALUE}.
```

**Comment by aadarsh-mohta** [comment]
Hi guys. We are using containerd in our environment with log driver as journald. We are seeing one issue where with containerd restart a lot of our containers which are being run via nerdctl are restarting. In logs we see the following:
`containerd[2742376]: time="2025-08-05T06:15:39.403104794Z" level=warning msg="failed to send SIGTERM signal, killing logging shim" error="os: process already finished"
containerd[2742376]: time="2025-08-05T06:15:39.416547484Z" level=info msg="shim disconnected" id=7467fa1e6911584a37b3984561742e10bafaadb628e2a4f69ed630beddc567fe
containerd[2742376]: time="2025-08-05T06:15:39.416834544Z" level=warning msg="cleaning up after shim disconnected" id=7467fa1e6911584a37b3984561742e10bafaadb628e2a4f69ed630beddc567fe namespace=default
containerd[2742376]: time="2025-08-05T06:15:39.416945776Z" level=info msg="cleaning up dead shim"`

Can someone suggest as to if this is expected. Ideally I would not want containers to restart with restart of containerd. What can I do to avoid the same ?
@fahedouch

**Comment by ningmingxiao** [comment]
//assign me @aadarsh-mohta   I will fix it later.
